// dom101/index.js
/*
 * this file is only provided for convenience and for tests.
 * it's not advised to use it.
 */

module.exports = {
  outerWidth: require('./outer-width'),
  outerHeight: require('./outer-height'),
  addClass: require('./add-class'),
  after: require('./after'),
  before: require('./before'),
  closest: require('./closest'),
  deepExtend: require('./deep-extend'),
  documentHeight: require('./document-height'),
  documentWidth: require('./document-width'),
  each: require('./each'),
  extend: require('./extend'),
  hasClass: require('./has-class'),
  isPlainObject: require('./is-plain-object'),
  map: require('./map'),
  matches: require('./matches'),
  on: require('./on'),
  prepend: require('./prepend'),
  querySelector: require('./query-selector'),
  querySelectorAll: require('./query-selector-all'),
  ready: require('./ready'),
  remove: require('./remove'),
  removeClass: require('./remove-class'),
  scrollTop: require('./scroll-top'),
  text: require('./text'),
  toggleClass: require('./toggle-class'),
  trigger: require('./trigger')
};
